### PR TITLE
patch: samd21: Correct missing underscore in defines for variants

### PR DESCRIPTION
**Pull request by rocky134**
Propagate fix for atsamd21j18a in a1a4c018074727620799ac526f4a26ce7b518ba3 to all samd21 variants.

### CODE REVIEW

**Comment by erikkallen** [comment]
I have tested this commit and it works for me on atsamd21g17a thanks!

**Review by erikkallen — APPROVED** [review]
The change looks good, the change was only done for the samd2118a and I think theese files where simply forgotten

**Review by erikkallen — APPROVED** [review]
The change looks good, the change was only done for the samd2118a and I think theese files where simply forgotten

**Review by nandojve — APPROVED** [review]

**Comment by nandojve** [comment]
@rocky134 , you need add ZephyrRTOS PR that update hal_atmel.

File: asf/sam0/include/samd21/samd21e15a.h [hunk 1]
```diff
@@ -62,15 +62,15 @@ typedef volatile       uint32_t RwReg;   /**< Read-Write 32-bit register (volati
 typedef volatile       uint16_t RwReg16; /**< Read-Write 16-bit register (volatile unsigned int) */
 typedef volatile       uint8_t  RwReg8;  /**< Read-Write  8-bit register (volatile unsigned int) */
 #if !defined(_UL)
```
Comment (stephanosio) [inline asf/sam0/include/samd21/samd21e15a.h:64]: This should be changed to check for `__UL` as well.

File: asf/sam0/include/samd21/samd21e15a.h [hunk 2]
```diff
@@ -62,15 +62,15 @@ typedef volatile       uint32_t RwReg;   /**< Read-Write 32-bit register (volati
...
-#define _U(x)          x ## U            /**< C code: Unsigned integer literal constant value */
-#define _L(x)          x ## L            /**< C code: Long integer literal constant value */
-#define _UL(x)         x ## UL           /**< C code: Unsigned Long integer literal constant value */
+#define __U(x)          x ## U            /**< C code: Unsigned integer literal constant value */
+#define __L(x)          x ## L            /**< C code: Long integer literal constant value */
+#define __UL(x)         x ## UL           /**< C code: Unsigned Long integer literal constant value */
 #endif
 #else
 #if !defined(_UL)
```
Comment (stephanosio) [inline asf/sam0/include/samd21/samd21e15a.h:70]: ditto

**Review by stephanosio — CHANGES_REQUESTED** [review]
Please add a commit message and sign-off as required by the project.

File: asf/sam0/include/samd21/samd21e15a.h (hunk 1 above)
Comment (rocky134) [reply asf/sam0/include/samd21/samd21e15a.h]: updated in all header files that were changed

File: asf/sam0/include/samd21/samd21e15a.h (hunk 2 above)
Comment (rocky134) [reply asf/sam0/include/samd21/samd21e15a.h]: updated in all header files that were changed

**Comment by rocky134** [comment]
> @rocky134 , you need add ZephyrRTOS PR that update hal_atmel.

No problem.
I will create a PR once this is resolved and merged.

**Review by stephanosio — COMMENTED** [review]
This PR, in itself, looks ok.

> I will create a PR once this is resolved and merged.

There must be a PR on the Zephyr-side that pulls this in, before this can be merged.

See https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules.

**Comment by rocky134** [comment]
> There must be a PR on the Zephyr-side that pulls this in, before this can be merged.

The pull request has been created
https://github.com/zephyrproject-rtos/zephyr/pull/42493

**Comment by nandojve** [comment]
Hi @stephanosio , Zephyr side is green. Is there anything pending from your side?

**Review by stephanosio — APPROVED** [review]

**Comment by stephanosio** [comment]
> Hi @stephanosio , Zephyr side is green. Is there anything pending from your side?

@nandojve nope, looks good to me.

**Comment by nandojve** [comment]
I started to get problems after this:

```
-- Zephyr version: 3.0.99 (/home/gfbudke/zephyr/mainline/zephyr), build: zephyr-v3.0.0-942-g0e5d438ee2ab
[135/173] Building C object zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/flash_sam0.c.obj
In file included from /home/gfbudke/zephyr/mainline/zephyr/soc/arm/atmel_sam0/samd21/soc.h:34,
                 from /home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c:17:
/home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c: In function 'flash_sam0_write_protection':
/home/gfbudke/zephyr/mainline/modules/hal/atmel/asf/sam0/include/samd21/samd21g18a.h:535:31: warning: implicit declaration of function '_UL' [-Wimplicit-function-declaration]
  535 | #define FLASH_SIZE            _UL(0x00040000) /* 256 kB */
      |                               ^~~
/home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c:36:27: note: in expansion of macro 'FLASH_SIZE'
   36 | #define LOCK_REGION_SIZE (FLASH_SIZE / LOCK_REGIONS)
      |                           ^~~~~~~~~~
/home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c:394:17: note: in expansion of macro 'LOCK_REGION_SIZE'
  394 |       offset += LOCK_REGION_SIZE) {
      |                 ^~~~~~~~~~~~~~~~
[163/173] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/gfbudke/zephyr/mainline/zephyr/build/zephyr/zephyr_pre0.map 
```

What problem this change fix once all code have _UL or _UL_ inside it?

**Comment by rocky134** [comment]
> I started to get problems after this:
> 
> ```
> -- Zephyr version: 3.0.99 (/home/gfbudke/zephyr/mainline/zephyr), build: zephyr-v3.0.0-942-g0e5d438ee2ab
> [135/173] Building C object zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/flash_sam0.c.obj
> In file included from /home/gfbudke/zephyr/mainline/zephyr/soc/arm/atmel_sam0/samd21/soc.h:34,
>                  from /home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c:17:
> /home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c: In function 'flash_sam0_write_protection':
> /home/gfbudke/zephyr/mainline/modules/hal/atmel/asf/sam0/include/samd21/samd21g18a.h:535:31: warning: implicit declaration of function '_UL' [-Wimplicit-function-declaration]
>   535 | #define FLASH_SIZE            _UL(0x00040000) /* 256 kB */
>       |                               ^~~
> /home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c:36:27: note: in expansion of macro 'FLASH_SIZE'
>    36 | #define LOCK_REGION_SIZE (FLASH_SIZE / LOCK_REGIONS)
>       |                           ^~~~~~~~~~
> /home/gfbudke/zephyr/mainline/zephyr/drivers/flash/flash_sam0.c:394:17: note: in expansion of macro 'LOCK_REGION_SIZE'
>   394 |       offset += LOCK_REGION_SIZE) {
>       |                 ^~~~~~~~~~~~~~~~
> [163/173] Linking C executable zephyr/zephyr_pre0.elf
> FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/gfbudke/zephyr/mainline/zephyr/build/zephyr/zephyr_pre0.map 
> ```
> 
> What problem this change fix once all code have _UL or _UL_ inside it? Reverting I not see problem anymore.

I originally discovered this problem when I got the below when building for the atsamd21g17, a variant of the atsamd21j18a with less ram and peripherals but otherwise the same core. The error snippet was as follows.
```
.....
C:/zephyrproject/modules/hal/atmel/asf/sam0/include/samd21/component/sercom.h:1172:37: warning: implicit declaration of function '__U' [-Wimplicit-function-declaration]
 1172 | #define SERCOM_USART_SYNCBUSY_MASK  __U(0x00000007) /**< \brief (SERCOM_USART_SYNCBUSY) MASK Register */
      |                                     ^~~
C:/zephyrproject/zephyr/drivers/serial/uart_sam0.c:98:32: note: in expansion of macro 'SERCOM_USART_SYNCBUSY_MASK'
   98 |  while ((usart->SYNCBUSY.reg & SERCOM_USART_SYNCBUSY_MASK) != 0) {
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
[67/152] Building C object zephyr/drivers/watchdog/CMakeFiles/drivers__watchdog.dir/wdt_sam0.c.obj
In file included from C:/zephyrproject/modules/hal/atmel/asf/sam0/include/samd21/samd21g17a.h:262,
                 from C:/zephyrproject/zephyr/soc/arm/atmel_sam0/samd21/soc.h:32,
                 from C:/zephyrproject/zephyr/drivers/watchdog/wdt_sam0.c:10:
C:/zephyrproject/zephyr/drivers/watchdog/wdt_sam0.c: In function 'wdt_sam0_isr':
C:/zephyrproject/modules/hal/atmel/asf/sam0/include/samd21/component/wdt.h:230:38: warning: implicit declaration of function '__U' [-Wimplicit-function-declaration]
  230 | #define WDT_INTFLAG_EW              (__U(0x1) << WDT_INTFLAG_EW_Pos)
......
```
That repeated for several hundred lines in different header files with the same overarching error `__UL, __U, __L` defines were missing.
After searching this repo, I found https://github.com/zephyrproject-rtos/hal_atmel/commit/a1a4c018074727620799ac526f4a26ce7b518ba3 and saw that had the same change that when done one samd21g17a fixed the issue so the obvious conclusion was that the change was initially only done only for atsamd21j18a which is used on popular Arduino boards and not for less common variants usually only used when designing custom boards. With that in mind, I submitted this PR.

However, It seems after your issue that the bigger issue is the inconsistent use of `_UL` vs `__UL` in the asf headers.
after seeing this, perhaps a better fix would be to revert the change and add
```
#define __UL(x) _UL(x)
#define __L(x) _L(x)
#define __U(x) _U(x)
```
as a compatibility header, as a cmake compile define, or replace the two defines in every asf file into a single define for each `__UL/_UL, __U/_U, __L/_L` pair.

Note: this change was identified on kernel tag. v2.7.0-rc1 there might have been something that was changed since then. I have not used kernel versions later than this.

**Pull request merged**